quil-rs/src/program/mod.rs

```diff
@@ -348,6 +360,90 @@ impl Program {
         Ok(expanded_program)
     }
 
+    /// Resolve [`LabelPlaceholder`]s and [`QubitPlaceholder`]s within the program such that the resolved values
+    /// will remain unique to that placeholder within the scope of the program.
+    ///
+    /// If you provide `label_resolver` and/or `qubit_resolver`, those will be used to resolve those values respectively.
+    /// If your placeholder returns `None` for a particular placeholder, it will not be replaced but will be left as a placeholder.
+    ///
+    /// If you do not provide a resolver for a placeholder, a default resolver will be used which will generate a unique value
+    /// for that placeholder within the scope of the program using an auto-incrementing value (for qubit) or suffix (for label)
+    /// while ensuring that unique value is not already in use within the program.
+    #[allow(clippy::type_complexity)]
+    pub fn resolve_placeholders(
+        &mut self,
+        label_resolver: Option<Box<dyn Fn(&LabelPlaceholder) -> Option<String>>>,
+        qubit_resolver: Option<Box<dyn Fn(&QubitPlaceholder) -> Option<u64>>>,
```

Comment:
Thoughts on splitting this up into two methods?

```rs
resolve_placeholders(&mut self)

resolve_placeholders_with_resolvers(
    &mut self,
    label_resolver: Box<dyn Fn(&LabelPlaceholder)> -> Option<String>>>,
    qubit_resolver: Box<dyn Fn(&QubitPlaceholder)> -> Option<u64>>>)
)
```

Some benefits of this:

1. The default behavior (`resolve_placeholders(None, None)`) seems like it would be a  common way to use this method and splitting it allows it to be used without specifying two `None`s each time.
2. Eliminates optional parameters where `None` actually means use some default behavior.
3. Less likely, but the return type for `qubit_resolver` is `Option<u64>`, if we ever change the type for a fixed qubit we'd have to break the signature. If we split them out, then at least users using the "default" one won't be broken.

Reply:
Yep, on board with this, except that your new signature doesn't allow for one resolver to be provided without the other. Is that intended?

Reply:
My thought was that there could be a default implementation that a user could pull in and explicitly use instead of using `None` resolving to that default implementation. Could also just split this into another set of methods: `resolve_placeholders_with_qubit_resolver` and ``resolve_placeholders_with_label_resolver`

Reply:
Ah, I see - we could indeed vend those defaults outside of the function body. That makes sense. Then the user can mix and match and wrap.

Otherwise - from your note - do you think we should replace the `Option<u64>` return type with `Option<Qubit>`

Reply:
I think `u64` is _probably_ fine. I'm not sure if getting a placeholder for a variable qubit is particularly useful (pyQuil at least only supports replacing placheolder with indexed qubits). I wish we could restrict the type to `Qubit::Fixed` specifically, but `u64` seems unlikely to ever be incompatible with a fixed qubit. Changing the signature to `Option<Qubit>` opens up the user to replacing a placeholder with a placeholder, which would be invalid, so it comes with it's own problems.

Reply:
I added those defaults and removed the option from the signature.